versioned/storage/cache/src/main/java/org/projectnessie/versioned/storage/cache/DistributedInvalidationsCacheBackend.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2024 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.storage.cache;
+
+import jakarta.annotation.Nonnull;
+import org.projectnessie.versioned.storage.common.persist.Obj;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+import org.projectnessie.versioned.storage.common.persist.Persist;
+import org.projectnessie.versioned.storage.common.persist.Reference;
+import org.projectnessie.versioned.storage.common.persist.UpdateableObj;
+
+final class DistributedInvalidationsCacheBackend implements CacheBackend {
+  private final CacheBackend local;
+  private final DistributedCacheInvalidation sender;
+
+  DistributedInvalidationsCacheBackend(
+      DistributedCacheInvalidations distributedCacheInvalidations) {
+    this.local = distributedCacheInvalidations.localBackend();
+    this.sender = distributedCacheInvalidations.invalidationSender();
+    distributedCacheInvalidations
+        .invalidationListenerReceiver()
+        .accept(
+            new DistributedCacheInvalidation() {
+              @Override
+              public void removeObj(String repositoryId, ObjId objId) {
+                local.remove(repositoryId, objId);
+              }
+
+              @Override
+              public void putObj(String repositoryId, ObjId objId, int hash) {
+                Obj existing = local.get(repositoryId, objId);
+                if (existing != null && existing.hashCode() != hash) {
+                  local.remove(repositoryId, objId);
```

Comment:
Java object hash collisions are permitted. Would it make sense to remove unconditionally here?

Reply:
It must/should not remove the object if it has changed in the meantime.

Reply:
Good point - removed the hash-stuff entirely